### PR TITLE
8187309: TreeCell must not change tree's data

### PR DESCRIPTION
**Pull request by kleopatra**
Issue was TreeView commit editing implementation violated the spec'ed mechanism:

- no default commit handler on TreeView
- TreeCell modifying the data directly

Fix is to move the saving of the edited value from cell into a default commit handler in tree and set that handler in the constructor.

Added tests that failed/passed before/after the fix (along with a sanity test for default commit that passed also before). Also fixed a test bug (incorrect registration of custom commit handler, see [JDK-8280951)](https://bugs.openjdk.java.net/browse/JDK-8280951) in TreeViewTest.test_rt_29650 to keep it passing.

### CODE REVIEW

**Review by mstr2 — APPROVED** [review]
Looks good to me.

File: modules/javafx.controls/src/test/java/test/javafx/scene/control/TreeCellTest.java [hunk 1]
```diff
@@ -888,6 +889,85 @@ public void testCommitEditMustNotFireCancel() {
...
+        String oldValue = editingItem.getValue();
+        // do nothing handler
+        tree.setOnEditCommit(e -> {});
+        tree.edit(editingItem);
+        String value = "edited";
+        cell.commitEdit(value);
+        assertEquals("cell text must not have changed", oldValue, cell.getText());
+    }
+
+
```
Comment (Maran23) [inline modules/javafx.controls/src/test/java/test/javafx/scene/control/TreeCellTest.java:944]: Very minor: Empty line

File: modules/javafx.controls/src/test/java/test/javafx/scene/control/TreeCellTest.java [hunk 2]
```diff
@@ -888,6 +889,85 @@ public void testCommitEditMustNotFireCancel() {
...
+     */
+    private TreeItem<String> setupForEditing(TreeCell<String> editingCell) {
+        tree.setEditable(true);
+        editingCell.updateTreeView(tree);
+        editingCell.updateIndex(1);
+        return editingCell.getTreeItem();
+    }
+
+    /**
+     * Test test setup.
```
Comment (Maran23) [inline modules/javafx.controls/src/test/java/test/javafx/scene/control/TreeCellTest.java:957]: Minor: I would rephrase that a bit. Something like:
`Tests the cell editing setup`.

**Review by Maran23 — APPROVED** [review]
Looks good to me too. The commitEdit(..) method of TreeCell is now in sync with the other cells. 
Just left two minor comments.

File: modules/javafx.controls/src/main/java/javafx/scene/control/TreeView.java [hunk 3]
```diff
@@ -337,6 +337,8 @@ public TreeView(TreeItem<T> root) {
         MultipleSelectionModel<TreeItem<T>> sm = new TreeViewBitSetSelectionModel<T>(this);
         setSelectionModel(sm);
         setFocusModel(new TreeViewFocusModel<T>(this));
+
+        setOnEditCommit(DEFAULT_EDIT_COMMIT_HANDLER);
```
Comment (aghaisas) [inline modules/javafx.controls/src/main/java/javafx/scene/control/TreeView.java:341]: You are adding the default edit commit handler to TreeView. Although it seems to be correct, this default handler is likely to be overwritten if the user code already has a setOnEditCommit() call. This is shown by the test_rt_29650() failure in TreeviewTest.java which you have corrected.

The changes done in this PR makes TreeView similar to ListView and TableView in terms of having the default edit commit handler.

Unfortunately, I do not see how can we avoid user code accidentally overwriting the default edit commit handler. Documenting this possibility seems to be the only way ahead.
Any other idea?

Comment (aghaisas) [reply modules/javafx.controls/src/main/java/javafx/scene/control/TreeView.java:341]: Thanks for the explanation. I see that it is already well documented.

**Review by aghaisas — APPROVED** [review]

File: modules/javafx.controls/src/test/java/test/javafx/scene/control/TreeCellTest.java (hunk 2 above)
Comment (kleopatra) [reply modules/javafx.controls/src/test/java/test/javafx/scene/control/TreeCellTest.java:957]: oops .. forgot to change this (and the other comment) before integration - sry
